File: functions/go/internal/premium/premium_test.go
package premium

import (
	"testing"
	"time"
)

func TestIsEffectivePremiumUsesSubscriptionWhileTierCatchesUp(t *testing.T) {
	now := time.Date(2026, 9, 13, 0, 0, 0, 0, time.UTC)
	cases := []struct {
		name string
		data map[string]any
		want bool
	}{
		{
			name: "tierはfreeでも有効期限内",
			data: map[string]any{"tier": "free", "subscription": map[string]any{
				"status": "active", "expires_at": now.Add(time.Hour),
			}},
			want: true,
		},
		{
			name: "tierはfreeでも買い切り",
			data: map[string]any{"tier": "free", "subscription": map[string]any{
				"lifetime": true, "status": "active",
			}},
			want: true,
		},
		{
			name: "取消済み買い切りは復活させない",
			data: map[string]any{"tier": "free", "subscription": map[string]any{
				"lifetime": true, "status": "expired",
			}},
			want: false,
		},
		{
			name: "取消済み購読は期限が未来でも復活させない",
			data: map[string]any{"tier": "free", "subscription": map[string]any{
				"status": "expired", "expires_at": now.Add(time.Hour),
			}},
			want: false,
		},
		{
			name: "保留中の購読は期限が未来でも復活させない",
			data: map[string]any{"tier": "free", "subscription": map[string]any{
				"status": "on_hold", "expires_at": now.Add(time.Hour),
			}},
			want: false,
		},
		{
			name: "猶予期間内",
			data: map[string]any{"tier": "free", "subscription": map[string]any{
				"status": "grace_period", "expires_at": now.Add(-24 * time.Hour),
			}},
			want: true,
		},
		{
			name: "期限切れ",
			data: map[string]any{"tier": "free", "subscription": map[string]any{
				"status": "expired", "expires_at": now.Add(-time.Hour),
			}},
			want: false,
		},
		{
			name: "期限の無い月額記録",
			data: map[string]any{"tier": "free", "subscription": map[string]any{
				"status": "active", "lifetime": false,
			}},
			want: false,
		},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := IsEffectivePremium(tc.data, now); got != tc.want {
				t.Fatalf("IsEffectivePremium() = %v, want %v", got, tc.want)
			}
		})
	}
}

File: functions/javascript/src/__tests__/premium.test.ts
import {isEffectivePremium} from '../utils/premium';

const timestamp = (ms: number) => ({toMillis: () => ms});

describe('isEffectivePremium', () => {
  const now = Date.UTC(2026, 8, 13);

  test('tier反映待ちでも有効期限内の購読をpremiumとして扱う', () => {
    expect(isEffectivePremium({
      tier: 'free',
      subscription: {
        status: 'active',
        expires_at: timestamp(now + 60_000),
      },
    }, now)).toBe(true);
  });

  test('期限切れの購読はpremiumとして扱わない', () => {
    expect(isEffectivePremium({
      tier: 'free',
      subscription: {
        status: 'expired',
        expires_at: timestamp(now - 60_000),
      },
    }, now)).toBe(false);
  });

  test('tier反映待ちでも買い切りはpremiumとして扱う', () => {
    expect(isEffectivePremium({
      tier: 'free',
      subscription: {lifetime: true, status: 'active'},
    }, now)).toBe(true);
  });

  test('取消済み買い切りをpremiumへ復活させない', () => {
    expect(isEffectivePremium({
      tier: 'free',
      subscription: {lifetime: true, status: 'expired'},
    }, now)).toBe(false);
  });

  test('取消済み購読は期限が未来でもpremiumへ復活させない', () => {
    expect(isEffectivePremium({
      tier: 'free',
      subscription: {
        status: 'expired',
        expires_at: timestamp(now + 60_000),
      },
    }, now)).toBe(false);
  });

  test('保留中の購読は期限が未来でもpremiumへ復活させない', () => {
    expect(isEffectivePremium({
      tier: 'free',
      subscription: {
        status: 'on_hold',
        expires_at: timestamp(now + 60_000),
      },
    }, now)).toBe(false);
  });
});
